### ISSUE TITLE
Feature Request: Add the ability to add context to an event on non OK statuses

### ISSUE THREAD

**Issue by majormoses** [sensu/sensu#1680]
## Expected Behavior
I have often felt sensu makes it hard to provide meaningful context on alerts. I have thought about this and think that we need to support adding additional context outside of check logic. To better explain I am going to give an example and then outline some use cases.

When monitoring something like high CPU % it would be helpful to before sending the alert to say include the top 10 processes by CPU. In this case its not as simple as just include a link to a graph as most people graph overall CPU usage and maybe certain processes but not for all processes on the system. 

Another Example (added based on discussions below) would be healthchecking a service over http, if the service is returning a non ok then we might want to add context such as heap usage if its a java application.

Here is my proposal as to the types of use cases and how a user might set this up.

in check:
```json
{
  "context": {
    "type": "command",
    "retrieve": "some_command_here"
  }
}
```

I think we should support at least 3 types:
- command
- link
- arbitrary ruby

## Current Behavior
In order to do this every check must be responsible for providing its own relevant context (which is pretty insane when you consider all the use cases for say check-http....

## Possible Solution
Enhance sensu to allow adding this extra context.

## Context
How appropriate you ask for this...

**Comment by mcbsdEU** [comment]
In our environment, we prefer to use our scripts in the check definitions. Community plugins are nice, but have lack of flexibility when adjustment is needed as every env. is often noticeably unique.
I am just thinking, that I would enhance the script so after detecting W|C threshold it would print top 10 processes in addition to common output like "CPU has reached....". This combined std out is automatically visible in ticketing system to which we sent events.
From my perspective, meaningful context is as much meaningful as you can make it.
But I am not saying that there are not use cases where this feature would have visible benefits...just can't think of them right now.
My 5 cents...cheers.

**Comment by majormoses** [comment]
Agreed that when there is clear context it makes sense to put it in the check directly but there are cases where the context is entirely independent of check logic. For example check-http.rb when its failing it might be helpful to include say heap information if its a java app. Rather than creating a bunch of different almost identical scripts just to add different context seems less desirable than adding the ability to allow this to be defined in the check definition.

**Comment by mcbsdEU** [comment]
I see...
Just thinking loud...I would extend the script and add options like "--heap"  when heap info is needed. You end up with same script, just different calls. We have modifications like that if we want to force the script to use ipv6 instead of v4 etc.
Depends on complexity...

**Comment by majormoses** [comment]
If you have a limited number of use cases that works but when writing generic monitors (like many of the ~200 community plugins which I help maintain) that leads to a ton of flags (thinking back to check-http.rb and in general community plugins there are a lot things that might be appropriate context so there could be that many flags in each check script) and lots of extra complexity in each check rather than simply having a callback to just append this in a much cleaner fashion.

**Comment by mcbsdEU** [comment]
I agree on this...

**Comment by portertech** [comment]
Closed by #1748 👍
